extensions/amp-render/1.0/amp-render.js

```diff
@@ -254,45 +266,44 @@ export class AmpRender extends BaseElement {
         dict({
           'render': (data) => {
             if (this.element.getAttribute('binding') === 'no') {
-              return templates
-                .renderTemplateAsString(dev().assertElement(template), data)
-                .then((html) => dict({'__html': html}));
+              return this.renderTemplateAsString_(data);
             }
 
-            let element;
-            return templates
-              .renderTemplate(dev().assertElement(template), data)
-              .then((el) => {
-                element = el;
-                return Services.bindForDocOrNull(this.element);
-              })
-              .then((bind) => {
+            let bind, element;
+            return Services.bindForDocOrNull(this.element)
+              .then((b) => {
+                bind = b;
                 if (!bind) {
-                  throw new Error();
+                  throw new Error('bind unavailable');
```

Comment:
This shouldn't be an error, since it would be unavailable when unused.

Reply:
Throwing an error to trigger the catch block where we render template as string. Do u have another preference?

Reply:
The runtime already throws a "not installed" error when it's not imported on documents that use `[*]` attributes - [example](https://playground.amp.dev/#share=PCFET0NUWVBFIGh0bWw+CjxodG1sIOKaoT4KICA8aGVhZD4KICAgIDxtZXRhIGNoYXJzZXQ9InV0Zi04IiAvPgogICAgPHRpdGxlPk15IEFNUCBQYWdlPC90aXRsZT4KICAgIDxsaW5rIHJlbD0iY2Fub25pY2FsIiBocmVmPSJzZWxmLmh0bWwiIC8+CiAgICA8bWV0YSBuYW1lPSJ2aWV3cG9ydCIgY29udGVudD0id2lkdGg9ZGV2aWNlLXdpZHRoIiAvPgogICAgPHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1vei1hbmltYXRpb246LWFtcC1zdGFydCA4cyBzdGVwcygxLGVuZCkgMHMgMSBub3JtYWwgYm90aDstbXMtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7YW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGh9QC13ZWJraXQta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUAtbW96LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1zLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW8ta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUBrZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19PC9zdHlsZT48bm9zY3JpcHQ+PHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOm5vbmU7LW1vei1hbmltYXRpb246bm9uZTstbXMtYW5pbWF0aW9uOm5vbmU7YW5pbWF0aW9uOm5vbmV9PC9zdHlsZT48L25vc2NyaXB0PgogICAgPHNjcmlwdCBhc3luYyBzcmM9Imh0dHBzOi8vY2RuLmFtcHByb2plY3Qub3JnL3YwLmpzIj48L3NjcmlwdD4KICAgIDxzdHlsZSBhbXAtY3VzdG9tPgogICAgICBoMSB7CiAgICAgICAgbWFyZ2luOiAxcmVtOwogICAgICB9CiAgICA8L3N0eWxlPgogIDwvaGVhZD4KICA8Ym9keT4KICAgIDxoMSBbdGV4dF09Im5ld1RleHQiPkhlbGxvIEFNUEhUTUwgV29ybGQhPC9oMT4KICAgIDxidXR0b24gb249InRhcDpBTVAuc2V0U3RhdGUoe25ld1RleHQ6ICdmb28nfSkiPgogICAgICBjaGFuZ2UgdGV4dAogICAgPC9idXR0b24+CiAgPC9ib2R5Pgo8L2h0bWw+Cg==)
![image](https://user-images.githubusercontent.com/10456171/119731251-b47bd500-be44-11eb-9c63-a5602643cfe7.png)

Reply:
Done with some code refactoring.